extern/storage-sealing/input.go

```diff
@@ -311,41 +311,45 @@ func (m *Sealing) SectorAddPieceToAny(ctx context.Context, size abi.UnpaddedPiec
 	}
 
 	m.inputLk.Lock()
-	if _, exist := m.pendingPieces[proposalCID(deal)]; exist {
+	if pp, exist := m.pendingPieces[proposalCID(deal)]; exist {
```

Comment:
So there is one somewhat big issue with this - when a market process calls AddPiece, it passes a reader to the call, and that data will be streamed directly into the sector.

If the market process restarts in the middle, or when we already we have a waiting AddPiece call, the stream will be broken, so here we will be waiting for as AddPiece call which is guaranteed to fail - instead, I think we should:
* Wait like we do now;
* If we see an error, try to add the piece to `m.pendingPieces` again  - we know that our reader was not read from, so all the bytes are in correct places, and the stream can be handed off to workers

Tldr I'd make this if into a for loop trying to add the piece to `m.pendingPieces` as long as:
* After waiting we somehow see no error (which is unlikely if we're here, but can happen with weird timing)
* Context is cancelled
* The previous call errored, so we can now add our piece to `m.pendingPieces` again